front/pages/api/w/[wId]/data_sources/index.ts

```diff
@@ -172,6 +172,20 @@ async function handler(
         });
       }
 
+      let vault = null;
+      if (typeof req.body.vaultId === "string") {
+        vault = await VaultResource.fetchById(auth, req.body.vaultId);
+        if (!vault) {
+          return apiError(req, res, {
+            status_code: 404,
+            api_error: {
+              type: "vault_not_found",
+              message: "The vault you requested was not found.",
+            },
+          });
+        }
+      }
+
       const globalVault = await VaultResource.fetchWorkspaceGlobalVault(auth);
```

Comment:
no need to fetch globalVault if vaultId is passed
In all cases, vault or globalVault, I think we need to add a permission check here - `  auth.hasPermission([vault.acl()], 'write')`